### ISSUE TITLE
Tunnel launch fails on device with iOS 14

### ISSUE THREAD

**Issue by francybiga** [Subito-it/SBTUITestTunnel#106]
Context:
* iOS 14 requires to add the `NSLocalNetworkUsageDescription` and `NSBonjourService` keys to the app plist in order to use Bonjour (which is used to setup the Tunnel client / server connection. This is enforced when running the app on a device.

* The bonjour service can be correctly resolved when performing the resolution in the “main app” (instead of xctestrunner app) after adding the needed key / values to the plist
* The bonjour service can be correctly resolved in xctestrunnerapp when executing on the simulator (the presence of key / values in only enforced on the device)

Problem:
* I add the required keys to the xctestrunner app Info.plist file but, nonetheless, I still fail to resolve Bonjour service, when running on a device, with the following error dictionary 

`["NSNetServicesErrorCode": -72008, "NSNetServicesErrorDomain": 10]`

Where the error code corresponds to `NSNetServicesMissingRequiredConfigurationError`, which indicates that the required key  / values are not in the plist. 

I found out that the xcrunner app actually has 2 different InfoPlist files:
+ one  inside the `APPNAMEUITests-Runner.app` bundle which contains several key/values automatically added by the framework (several of type “UsageDescription” with value "Access is necessary for automated testing.”)
* another one in `APPNAMEUITestsE-Runner.app/PlugIns/APPNAMEUITests.xctest` has the contents of the `Info.plist` file I have inside the Xcode project and where I added the Bonjour key/values

I think that iOS expects the keys to be inside the first plist, which I don’t know how to edit inside my project (I suppose it’s completely generated by `XCTest`).

What’s the correct way to add custom key/values to the “main" xctestrunner plist? (in this case the key / values needed to use Bonjour on iOS 14) 

I opened both a [StackOverflow question](https://stackoverflow.com/questions/64880305/xctest-add-custom-key-value-to-the-xcuitest-runner-apps-info-plist) and a radar to Apple, but I decided to also open an issue here for reference in case some else has the same problem.

**Comment by tcamin** [comment]
Hi! Thanks for the detailed explanation, indeed there seems to be an issue with bonjour on physical devices.

As a proof of concept I tried to come up with a different approach that did not require to use bonjour which I pushed here https://github.com/Subito-it/SBTUITestTunnel/compare/feature/remove_bonjour

In the previous implementation the port on which to instantiate the tunnel was assigned by bonjour on the server side. In the new a random free port is chosen client side and injected to the server.

It "works" and it might even be faster than the previous implementation, however it suffers from 2 potential problems/race conditions.

1. It takes time for the aut to fire up, hence it may occur that the port that was chosen by the client might no longer be free
2. If multiple simulators (which share the same ip) start in parallel it could occur that different clients pick the same port

While I kind of like the simplicity of the new approach it might be safer to keep bonjour on simulators and use the "new" solution only on devices (which additionally don't suffer from problem 2.)

Give it a try and let me know what you think about it. In the next days I'll try to remerge bonjour in this branch adding conditionals for device/simulator

**Comment by tcamin** [comment]
Bonjour discovery restored for simulators

**Comment by francybiga** [comment]
I saw your solution. I agree to use bonjour when running on simulators to avoid potential port collisions, and let the client choose the port without using bonjour when running on devices. 👍

**Comment by whizbuff** [comment]
Hi,
Is proposed solution merged ?

**Comment by tcamin** [comment]
I only had the chance to do some brief testing but running this new implementation against our large test suite proved to be less reliable than the current one. I wasn't able to dig into more details but apparently there were more test failures due to issues establishing the tunnel. This task is definitely important and I'll try to clean things up as soon as I can. In the meantime you could just update the Podfile and pin it to this branch

**Comment by whizbuff** [comment]
thank you for response and suggestion.
keep up the great work 👍

**Comment by ThibaudSE** [comment]
Hello, any update regarding this issue ?

**Comment by tcamin** [comment]
The current implementation seems to be working, you can definitely start using it by pinning to this branch. I did not have time to validate the changes against our setup which is mandatory before merging it into master. Sorry about the delay

**Comment by elneruda** [comment]
Hello @tcamin do you plan to merge this [branch](https://github.com/Subito-it/SBTUITestTunnel/compare/feature/remove_bonjour) soon ?

**Comment by tcamin** [comment]
@elneruda unfortunately we've been hitting some rare issues when running this branch on our test suite with the number of failed connections slightly increasing when compared to the stable implementation. Unfortunately I did not have time to investigate this further. 

My plan would be to retest everything with Xcode 13/iOS 15 to see if there are other additional regressions.

**Comment by jeremygrosjean** [comment]
Hey @tcamin,
Thanks for your work!
Right now we are using the branch in our test. But it seems that there is more and more random failed connection as you said previously. Do you have any idea of something we can do to reduce it?

**Comment by tcamin** [comment]
Unfortunately I did not have time to dig into debugging that. That’s one of the reasons I still did not merge this branch in. If you manage to replicate the failed connections in a systematic way please let me know

**Comment by msalmanST** [comment]
Hi Guys,

I am also facing an issue while launching tunnel (it is happening on simulator). On line app.launchTunnel() it is throwing "Failed waiting for app to be ready (NSInternalInconsistencyException)"

My lib version is 6.6.4.

**Comment by shoaibsadanaut** [comment]
@msalmanST I am facing same issue now, It was working fine previously but suddenly stopped working. Any solution?

**Comment by msalmanST** [comment]
I couldn't solve this issue and hence moved from this lib, instead I am using alamofire testing feature to add mock api in my app.

**Comment by shoaibsadanaut** [comment]
@msalmanST I am afraid I can't do that because my test suit is coupled with SBTUITestTunnel, one example is **performCustomCommandNamed** setting dynamic values from test runner to app. 
Do you know any other solution where I can do this?

**Comment by msalmanST** [comment]
I doubt you might have to take a long route or may be look for some other lib.

Also in xcode 13 there is a retry test mechanism introduce, so you can take advantage of that as well, i,e retry x no of times to make sure that test run.

**Comment by tcamin** [comment]
@shoaibsadanaut a few questions:

- Did you try running the tests in the SBTUITestTunnel's sample project?
- What version of macOS are you running? There has been a severe regression in macOS 11.3 and newer (see https://developer.apple.com/forums/thread/679098 and https://twitter.com/tomascamin/status/1387315074185011203?s=20) which impacts the app startup
- How big is the project you're integrating SBTUITestTunnel?

**Comment by shoaibsadanaut** [comment]
@tcamin 

- We have no problem running it on local machine only our CI is failing and we are using Bitrise for that 
- Please have a look at image stack we are using on CI
- Project is big, it is a fin-tech application. It is not small or medium scale project. 

<img width="647" alt="Screenshot 2021-09-10 at 6 27 26 PM" src="https://user-images.githubusercontent.com/73075608/132860657-977adb30-04fe-4c19-9f35-789134490675.png">

**Comment by shoaibsadanaut** [comment]
@tcamin It's crashing or due to bonjour I am not sure. 

<img width="861" alt="Screenshot 2021-09-10 at 6 42 55 PM" src="https://user-images.githubusercontent.com/73075608/132862869-f94746ac-ca46-4985-81f5-2aa2c6b5e8e6.png">

**Comment by tcamin** [comment]
Did you try disabling the debugger (uncheck _Debug executable_)  in the scheme both for Run and Test?

<img width="933" alt="Screenshot 2021-09-10 at 15 52 24" src="https://user-images.githubusercontent.com/875025/132864093-757ae3cf-d0f0-4a54-9c74-e5c90374127a.png">

**Comment by shoaibsadanaut** [comment]
@tcamin yes and the result is same.

**Comment by tcamin** [comment]
Did it ever worked on Bitrise?

**Comment by shoaibsadanaut** [comment]
@tcamin Yes it was working since last six months and yesterday it stopped working. We haven’t changed anything from our side.

**Comment by shoaibsadanaut** [comment]
No we are using 6.6.4 from the start.

**Comment by tcamin** [comment]
If nothing changed I'm unsure it's something strictly related to the tunnel. Did you double check that the vm is still really on 10.15.7?

**Comment by shoaibsadanaut** [comment]
@tcamin Yes it 10.5.7. I tried couple of stacks even with Big Sur and app did no pass the splash screen, for me it is confusing because we have not changed anything UI tests related. 

![Screenshot 2021-09-11 at 9 30 01 PM](https://user-images.githubusercontent.com/73075608/132954741-196a33e6-1140-4b73-9dc1-d554d4a604ca.png)

**Comment by tcamin** [comment]
Did you try bisecting changes starting from the last working job?

**Comment by shoaibsadanaut** [comment]
@tcamin yep but not working.

**Comment by tcamin** [comment]
If old runs that were ok no longer work, I'd say something changed in the CI setup/configuration.

Did you try running a simple sample app with the tunnel to see if that works?

**Comment by shoaibsadanaut** [comment]
Will try that, query sent to bitrise as well.

**Comment by tcamin** [comment]
Please reply back if you find out more. Thanks

**Comment by shoaibsadanaut** [comment]
Hey man sorry to ping you but do you know what can cause this error? @tcamin 

 ✗ testCoreUserJourney, [SBTUITestTunnel] Failed getting IPC proxy, Error Domain=DTXIPCErrorDomain Code=1 "Connection <DTXIPCConnection: 0x600002ee1b00> is invalid." UserInfo={NSLocalizedDescription=Connection <DTXIPCConnection: 0x600002ee1b00> is invalid.} (NSInternalInconsistencyException)

**Comment by shoaibsadanaut** [comment]
Thanks for answering my queries. I have one last question
**NSMakeSpecialForwardingCaptureBlock** is being used for Interprocess communication? The tunnel is linked with my QA build and the app store is warning me about non-public API usage. ITMS-90338: Non-public API usage.
@tcamin

**Comment by shoaibsadanaut** [comment]
@tcamin it works fine in 8.1.0 no non-public API warning.

**Comment by pochikaevams** [comment]
@shoaibsadanaut Hello, did you find what was wrong with this error?  "Failed getting IPC proxy" - i am facing the same error.

**Comment by jwu-gusto** [comment]
same here. I get this error on bitrise.
